numba/_dictobject.c

```diff
@@ -0,0 +1,1079 @@
+a combined table, then the me_value slots in the old table are NULLed out.
+After resizing a table is always combined,
+but can be resplit by make_keys_shared().
+*/
+int
+numba_dict_resize(NB_Dict *d, Py_ssize_t minsize) {
+    Py_ssize_t newsize, numentries;
+    NB_DictKeys *oldkeys;
+    int status;
+
+    /* Find the smallest table size > minused. */
+    for (newsize = D_MINSIZE;
+         newsize < minsize && newsize > 0;
+         newsize <<= 1)
+        ;
+    if (newsize <= 0) {
+        return ERR_NO_MEMORY;
+    }
+    oldkeys = d->keys;
+
+    /* NOTE: Current odict checks mp->ma_keys to detect resize happen.
+     * So we can't reuse oldkeys even if oldkeys->dk_size == newsize.
+     * TODO: Try reusing oldkeys when reimplement odict.
+     */
+
+    /* Allocate a new table. */
+    status = numba_dictkeys_new(
+        &d->keys, newsize, oldkeys->key_size, oldkeys->val_size
+    );
+    if (status != OK) {
+        d->keys = oldkeys;
+        return status;
+    }
+    // New table must be large enough.
+    assert(d->keys->usable >= d->used);
+
+    numentries = d->used;
+
+    if (oldkeys->nentries == numentries) {
+        NB_DictEntry *oldentries, *newentries;
+
+        oldentries = get_entry(oldkeys, 0);
+        newentries = get_entry(d->keys, 0);
+        memcpy(newentries, oldentries, numentries * oldkeys->entry_size);
+    }
+    else {
+        Py_ssize_t i;
+        size_t epi = 0;
+        for (i=0; i<numentries; ++i) {
+
+            /*
+                ep->hash == (-1) hash means it is empty
+
+                Here, we skip until a non empty entry is encountered.
+            */
+            while( get_entry(oldkeys, epi)->hash == DKIX_EMPTY ) {
+                assert( mem_cmp_zeros(entry_get_val(oldkeys, get_entry(oldkeys, epi)), oldkeys->val_size) == 0 );
+                epi += 1;
+            }
+            memcpy(
+                get_entry(d->keys, i),
+                get_entry(oldkeys, epi),
+                oldkeys->entry_size
+            );
+            epi += 1;
+
+        }
+
+    }
+    numba_dictkeys_free(oldkeys);
+
+    build_indices(d->keys, numentries);
+    d->keys->usable -= numentries;
+    d->keys->nentries = numentries;
+    return OK;
+}
+
+/*
+    Adapted from CPython delitem_common
+ */
+int
+numba_dict_delitem(NB_Dict *d, Py_hash_t hash, Py_ssize_t ix)
+{
+    Py_ssize_t hashpos;
+    NB_DictEntry *ep;
+    NB_DictKeys *dk = d->keys;
+
+    hashpos = lookdict_index(dk, hash, ix);
+    assert(hashpos >= 0);
+
+    d->used -= 1;
+    ep = get_entry(dk, ix);
+    set_index(dk, hashpos, DKIX_DUMMY);
+
+    zero_key(dk, entry_get_key(dk, ep));
+    zero_val(dk, entry_get_val(dk, ep));
+    ep->hash = DKIX_EMPTY; // to mark it as empty;
+
+    return OK;
+}
+
+
+/**
+ * Adapted from dict_popitem
+ *
+ */
+int
+numba_dict_popitem(NB_Dict *d, char *key_bytes, char *val_bytes)
+{
+    Py_ssize_t i, j;
+    char *key_ptr, *val_ptr;
+    NB_DictEntry *ep;
```

Comment:
The loop must execute if the dictionary is not malformed.  If `d->used == 0`, it will early exit without using `ep`.  Otherwise, the loop must have run at least once.  But, I will still put in the `ep = NULL` to ensure a malformed dictionary will trigger a segfault later by `ep->hash`.

Reply:
I think this is wise :)